### PR TITLE
Dynamic artifact handling

### PR DESCRIPTION
**Pull request by gladysteh99**
Refactor handling of artifact class to use entry points to prevent artifact module from blowing up.

### CODE REVIEW

File: lazyscribe/artifacts/__init__.py [hunk 1]
```diff
@@ -23,11 +24,30 @@ def _get_handler(alias: str) -> Type[Artifact]:
         The artifact handler class object. This object will need to be constructed
         using :py:meth:`lazyscribe.artifacts.Artifact.construct`.
     """
-    for obj in Artifact.__subclasses__():
-        if obj.alias == alias:
-            out = obj
+    eps = entry_points()
+    entry = eps.select(group="artifact_type")
```
Comment (ak-gupta) [inline lazyscribe/artifacts/__init__.py:28]: Hm, how do we feel about this entrypoint naming? IMO, something `lazyscribe` specific might be less likely to have a conflict with another package. Maybe ``"lazyscribe.artifact_types"``?

File: lazyscribe/artifacts/__init__.py [hunk 2]
```diff
@@ -23,11 +24,30 @@ def _get_handler(alias: str) -> Type[Artifact]:
...
+            mod, name = full_artifact_class.value.rsplit(".", 1)
+            try:
+                mod = import_module(mod)
+            except ImportError as imp:
+                raise RuntimeError(
+                    f"Unable to import handler for {alias} through entry points or standard import."
+                ) from imp
+
+            for part in name.split("."):
+                mod = getattr(mod, part)
```
Comment (ak-gupta) [inline lazyscribe/artifacts/__init__.py:41]: Using `rsplit` always feels a bit error prone (to me). I was playing around with this, what do you think? We can modify the `pyproject.toml` to specify `<path>.<to>:<HandlerClass>`:

```toml
[project.entry-points."artifact_type"]
base = "lazyscribe.artifacts.base:Artifact"
joblib = "lazyscribe.artifacts.joblib:JoblibArtifact"
json = "lazyscribe.artifacts.json:JSONArtifact"
testartifact = "tests.conftest:TestArtifact"
```

then we can use the `load()` method from entry points:

```suggestion
        try:
            mod = full_artifact_class.load()
        except ImportError as imp:
            raise RuntimeError(
                f"Unable to import handler for {alias} through entry points."
            ) from imp
```

I just did a quick look at it in Python 3.10, so not sure if it's a solution we can roll with fully.

File: lazyscribe/artifacts/__init__.py [hunk 3]
```diff
@@ -23,11 +24,30 @@ def _get_handler(alias: str) -> Type[Artifact]:
         The artifact handler class object. This object will need to be constructed
         using :py:meth:`lazyscribe.artifacts.Artifact.construct`.
     """
-    for obj in Artifact.__subclasses__():
-        if obj.alias == alias:
-            out = obj
```
Comment (ak-gupta) [inline lazyscribe/artifacts/__init__.py:28]: Do you think we should get rid of this pattern? I'm wondering whether it's helpful for users that are experimenting outside of the packaging context.

File: pyproject.toml [hunk 4]
```diff
@@ -37,6 +37,12 @@ repository = "https://github.com/lazyscribe/lazyscribe"
 requires = ["setuptools>=64.0.0"]
 build-backend = "setuptools.build_meta"
 
+[project.entry-points."artifact_type"]
+base = "lazyscribe.artifacts.base.Artifact"
+joblib = "lazyscribe.artifacts.joblib.JoblibArtifact"
+json = "lazyscribe.artifacts.json.JSONArtifact"
+testartifact = "tests.conftest.TestArtifact"
```
Comment (ak-gupta) [inline pyproject.toml:44]: Can we exclude the test-specific handler from being distributed with the package?

**Comment by ak-gupta** [comment]
Tagging additional reviewers since I contributed a couple small fixes and documentation.

**Review by jdawang — APPROVED** [review]

**Pull request merged**